rpc server: listen to `ipv6 socket` if available and `--experimental-rpc-endpoint` CLI option

niklasad1 opened the pull request:
Close https://github.com/paritytech/polkadot-sdk/issues/3488, https://github.com/paritytech/polkadot-sdk/issues/4331

This changes/adds the following:

1. The default setting is that substrate starts a rpc server that listens to localhost both Ipv4 and Ipv6 on the same port. Ipv6 is allowed to fail because some platforms may not support it
2. A new RPC CLI option `--experimental-rpc-endpoint` which allow to configure arbitrary listen addresses including the port, if this is enabled no other interfaces are enabled.
3. If the local addr is not found for any of the sockets the server is not started throws an error.
4. Remove the deny_unsafe from the RPC implementations instead this is an extension to allow different polices for different interfaces/sockets such one may enable unsafe on local interface and safe on only the external interface.

So for instance in this PR it's now possible to start up three RPC endpoints as follows:
```
$ polkadot --experimental-rpc-endpoint "listen-addr=127.0.0.1:9944,rpc-methods=unsafe" --experimental-rpc-endpoint "listen-addr=0.0.0.0:9945,rpc-methods=safe,rate-limit=100" --experimental-rpc-endpoint "listen-addr=[::1]:9944,optional=true"
```

#### Needs to be addressed

~1. Support binding to a random port if it's fails with the default stuff for backward compatible reasons~
~2. How to sync that the rpc CLI params and that the rpc-listen-addr align, hard to maintain...~
~3. Add similar warning prints for exposing unsafe methods on external interfaces..~
~4. Inline todos + the hacky String conversion from rpc params.~

#### Cons with this PR

Manual strings parsing impl more error-prone than relying on clap....

//cc @jsdw @BulatSaif @PierreBesson @bkchr

PierreBesson commented:
> make it possible to use different polices for each listening interface with a special notation which is very similar to an URL which is <ip:addr>/?rpc-methods=safe&cors=*

This is definitely very useful for multiple use cases and it is a feature that has been lacking for:
- Validator/Collator security hardening: having 2 listen addresses with different permissions: one for regular queries (eg. sync state check) and one for admin tasks (eg. setting up keys)
- RPC node: having 2 listen addresses, one public with rate limits and CORS and one internal which allows unsafe RPC calls.

jsdw commented:
> `$ polkadot --rpc-listen-addrs "127.0.0.1:9944:/?rpc-methods=unsafe, 0.0.0.0:9945:/?rpc-methods=safe&rpc-rate-limit=100"`

This format for allowing options to be passed with multiple listen addrs looks good to me!

BulatSaif commented:
Very useful feature. One remark: can we avoid using whitespace in the arguments? It sometimes causes issues. Ideally, allow specifying multiple `--rpc-listen-addr` arguments.

niklasad1 commented:
> Very useful feature. One remark: can we avoid using whitespace in the arguments? It sometimes causes issues. Ideally, allow specifying multiple --rpc-listen-addr arguments.

Yeah, it should without whitespaces but I can add a test for it.

pepoviola commented:
look great! just a `nit` comment about the naming. Since we are using the string to also set other options I think we can use something like `--rpc-listen-uri`. wdyt?

BulatSaif commented:
> Since we are using the string to also set other options I think we can use something like --rpc-listen-uri. wdyt?

If it is called a URI, it will impose some limitations on what can be included in the string, as it needs to follow [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986).

niklasad1 commented:
> look great! just a nit comment about the naming. Since we are using the string to also set other options I think we can use something like --rpc-listen-uri. wdyt?

The reason is that the format isn't strictly an URL/URI and currently the format `ip:port/<query str>` which isn't a valid URL because it's a relative URL without a base such as wss://127.0.0.1:9944. So I actually had to write some custom parsing for the query params but nothing complicated.

Thus, I think --rpc-listen-addrs is more accurate but open for other suggestions :)

pepoviola commented:
> > look great! just a nit comment about the naming. Since we are using the string to also set other options I think we can use something like --rpc-listen-uri. wdyt?
> 
> 
> 
> Yes, Bulat is correct
> 
> 
> 
> The reason is that the format isn't strictly an URL/URI and currently the format `ip:port/<query str>` which isn't a valid URL because it's a relative URL without a base such as `wss://127.0.0.1:9944`. So I actually had to write some custom parsing for the query params but nothing complicated.
> 
> 
> 
> Thus, I think --rpc-listen-addrs is more accurate but open for other suggestions :)

Then sounds good 👍.
Thx!!

skunert commented:
I like the direction. Overall, we have one million CLI args and it can be overwhelming, so it seems like a good idea to summarize them.

Should be stick to one format? If we are using this new string format for everything, why keep all the command line arguments around? IMO we should then deprecate and remove at some point.

I think currently the help message is relatively short. It says
```
      --rpc-listen-addrs <RPC_LISTEN_ADDRS>
          Specify the JSON-RPC server listen address.

          The format of the listen addr is `"<ip:port>/?<query params>"`

          For example: if you want to listen on `127.0.0.1:9944` and enable the `rpc-methods=unsafe query parameter, you can use: `--rpc_listen_addrs
          127.0.0.1:9933/?rpc-methods=unsafe&cors=*"`.
```
But how do I discover all the query parameters that are accepted? By looking at the other CLI flags and then translating?
We should probably list alll acceptable parameters.

About the format, I am not sure what I think about it. One alternative would be to just have kv pairs like:
```
polkadot --rpc-listen-addr "addr=127.0.0.1:9944,rpc-methods=unsafe" --rpc-listen-addr "addr=0.0.0.0:9945,rate-limit=100"
```
Not sure if that is better, but at least to me it seems more familiar, but not a strong opinion.

niklasad1 commented:
> Should be stick to one format? If we are using this new string format for everything, why keep all the command line arguments around? IMO we should then deprecate and remove at some point.

I agree but I haven't moved all rpc related params to this new format but that shouldn't require much code.

> But how do I discover all the query parameters that are accepted? By looking at the other CLI flags and then translating?
We should probably list alll acceptable parameters.

Yes good point, it's a bit annoying to describe everything manually but yeah it must be documented properly.

> About the format, I am not sure what I think about it. One alternative would be to just have kv pairs like:
```
polkadot --rpc-listen-addr "addr=127.0.0.1:9944,rpc-methods=unsafe" --rpc-listen-addr "addr=0.0.0.0:9945,rate-limit=100"
```
> Not sure if that is better, but at least to me it seems more familiar, but not a strong opinion.

I don't care that much either but I figured that everyone is familiar URL syntax but your suggestion is easier to read
without the query params syntax so I think I prefer it but lemme think about it.

niklasad1 commented on substrate/client/cli/src/commands/run_cmd.rs:
```diff
@@ -128,6 +128,43 @@ pub struct RunCmd {
+	///    (optional)
+	///  • max_batch_request_len: The maximum number of requests in a batch (optional)
+	///  • cors: The CORS allowed origins, this can enabled more than once (optional)
+	///  • rpc-methods: Which RPC methods to allow, valid values are "safe", "unsafe" and "auto"
```
```suggestion
	///  • methods: Which RPC methods to allow, valid values are "safe", "unsafe" and "auto"
```

niklasad1 commented:
1. So, I have addressed some your main feedback and changed `--rpc-listen-addrs` to `--rpc-endpoint` and moved to key-value pairs instead of the url-like syntax.
2. Then moved all rpc configurations to the RpcEndpoint struct, so it should be easy to deprecate the old CLI flags now but I would want
3. Documented all key-value pairs that may be configured.

As I see it the main downside with this is the manual parsing and the manual documentation but I reckon having this new scheme is still much better and that we will let folks specify the listen addr (this could be a footgun if mis-used) and enable any number of rpc endpoints.

Would be cool if you can have another look and I could add the deprecation stuff in this PR if you want.

skunert reviewed (commented):
Its still a bit of a shame that we need to have this custom format. But I see no better way of supporting multiple RPC-endpoints with different parameterization. 

We need to  make super sure that all the docs are consistent with the actual parameters, otherwise this will cause a lot of confusion.

Regarding deprecation, even if there is only one endpoint, IMO we should deprecate the other rpc commands and put a date of deprecation on it, maybe in 6 months. I know @kianenigma has thought about the  ergonomics of our node quite a bit, maybe he has an opinion.

---

One other thing I noticed while playing with this PR is that it breaks some zombienet tests.
e.g. running
```
zombienet -l text --provider native test cumulus/zombienet/tests/0003-full_node_catching_up.zndsl
```
Panics when using a polkadot binary build with on this branch. Apparently node eve can not connect to the RPC server on alice. Looking at alice it reports:
```
2024-08-22 11:51:38.662  INFO                 main sc_rpc_server: Running JSON-RPC server: addr=0.0.0.0:9944,[::]:39621    
```
Even though it runs with `--rpc-port 39971`, so maybe a bug was introduced here.

niklasad1 commented:
> Even though it runs with --rpc-port 39971, so maybe a bug was introduced here.

It's probably the parsing of the log:
` Running JSON-RPC server: addr=0.0.0.0:9944,[::]:39621`

Previously it was: 
`$ Running JSON-RPC server: addr=0.0.0.0:9944,allowed_origins=[*]`

So perhaps just it's just parsing the port that failed but I don't know....
Or do you know that zombienet actually connected to either 0.0.0.0:9944 or [::]:39621?  Then it's a bug but I don't think so.

So for instance in substrate/subxt we split does on which still works in the integration test etc

niklasad1 commented on substrate/client/cli/src/params/mod.rs:
```diff
@@ -160,9 +165,261 @@ pub struct NetworkSchemeFlag {
+
+		let batch_config = match (disable_batch_requests, max_batch_request_len) {
+			(Some(false), Some(_)) => {
+				return Err("`max-batch-request-len` cannot be specified when `disable-batch-requests` is false".to_string());
```
Those are mutually exclusive you can't both disable batch requests and set it a specific max limit on them. You are correct it should true.

The possible variants are:
- Disable batches
- Enable batch with no upper bound
- Limit batch some upper limit `n`, so if a batch with more than n entries it's just dropped but ACK:ed with an error.

niklasad1 replied:
if it's not clear I suggest to rename it or simplify it

skunert replied:
All good, I was just saying that we should check for `true` and not `false` here.

pepoviola commented:
> > Even though it runs with --rpc-port 39971, so maybe a bug was introduced here.
> 
> It's probably the parsing of the log: ` Running JSON-RPC server: addr=0.0.0.0:9944,[::]:39621`
> 
> Previously it was: `$ Running JSON-RPC server: addr=0.0.0.0:9944,allowed_origins=[*]`
> 
> So perhaps just it's just parsing the port that failed but I don't know.... Or do you know that zombienet actually connected to either 0.0.0.0:9944 or [::]:39621? Then it's a bug but I don't think so.
> 
> So for instance in substrate/subxt we split it based `addr=<sockaddr`, which still works in the integration test etc

Hi @niklasad1 / @skunert, let me try to reproduce since zombienet assign the ports without parse the logs (e.g with --rpc-port flag). 

Thx!

skunert commented:
> > Even though it runs with --rpc-port 39971, so maybe a bug was introduced here.
> 
> It's probably the parsing of the log: ` Running JSON-RPC server: addr=0.0.0.0:9944,[::]:39621`
> 
> Previously it was: `$ Running JSON-RPC server: addr=0.0.0.0:9944,allowed_origins=[*]`
> 
> So perhaps just it's just parsing the port that failed but I don't know.... Or do you know that zombienet actually connected to either 0.0.0.0:9944 or [::]:39621? Then it's a bug but I don't think so.
> 
> So for instance in substrate/subxt we split it based `addr=<sockaddr`, which still works in the integration test etc

I mean the problem is that I am specifying `--rpc-port 39971`, but the node says it listens to `Running JSON-RPC server: addr=0.0.0.0:9944,[::]:39621`, so different ports.

niklasad1 commented:
> I mean the problem is that I am specifying --rpc-port 39971, but the node says it listens to Running JSON-RPC server: addr=0.0.0.0:9944,[::]:39621, so different ports.

Aight, got it. Yepp, indeed it's a bug.

skunert commented on substrate/client/cli/src/commands/run_cmd.rs:
```diff
@@ -128,6 +128,43 @@ pub struct RunCmd {
+	///    accept (optional)
+	///  • max-request-size: The maximum size of a request body in megabytes (optional)
+	///  • max-response-size: The maximum size of a response body in megabytes (optional)
+	///  • max-subscriptions_per_connection: The maximum number of subscriptions per connection
```
```suggestion
	///  • max-subscriptions-per-connection: The maximum number of subscriptions per connection
```
Last one :D

niklasad1 replied:
oh sorry

skunert reviewed (approved):
LGTM

niklasad1 commented on substrate/client/cli/src/commands/run_cmd.rs:
```diff
@@ -128,6 +128,43 @@ pub struct RunCmd {
+	///  • max-batch-request-len: The maximum number of requests in a batch (optional)
+	///  • cors: The CORS allowed origins, this can enabled more than once (optional)
+	///  • methods: Which RPC methods to allow, valid values are "safe", "unsafe" and "auto"
+	///    (default)
```
```suggestion
	///    (optional)
```

niklasad1 commented on prdoc/pr_4792.prdoc:
```diff
@@ -0,0 +1,31 @@
+  - name: sc-rpc
+    bump: major
+  - name: sc-cli
+    bump: minor
```
```suggestion
    bump: minor
    
```

niklasad1 commented on prdoc/pr_4792.prdoc:
```diff
@@ -0,0 +1,32 @@
+
+      1. The default setting is that substrate starts a rpc server that listens to localhost both ipv4 and ipv6 on the same port. 
+         ipv6 is allowed to fail because some platforms may not support it
+      2. A new RPC CLI option `--rpc-endpoint` is introduced which allow to configure arbitrary listen addresses including the port, 
```
```suggestion
      2. A new RPC CLI option `--rpc-endpoint` is introduced which allows to configure arbitrary listen addresses including the port, 
```

niklasad1 commented on substrate/client/cli/src/commands/run_cmd.rs:
```diff
@@ -410,15 +447,68 @@ impl CliConfiguration for RunCmd {
+			for endpoint in &self.rpc_endpoint {
+				// Technically, `0.0.0.0` isn't a public IP address, but it's a way to listen on
+				// all interfaces. Thus, we consider it as a public endpoint and warn about it.
+				if endpoint.rpc_methods == RpcMethods::Unsafe && endpoint.is_global() ||
```
//cc @BulatSaif @PierreBesson I haven't introduced any flag such as `--unsafe-external` and `--external` here, it's up to the user to define such a thing by the listen-addr but I added a similar warning log.

Do you want extra flags for that anyway? 

Now one need to do:  `--rpc-endpoint "listen-addr=0.0.0.0:9944,methods=unsafe"

but we could add something extra such as option for

` --rpc-endpoint "listen-addr=0.0.0.0:9944,methods=unsafe,unsafe-external=true"` but I don't see how that could make things better though...

niklasad1 commented:
Another option is call this option `--rpc-endpoint-unstable` until we collect some feedback on this approach and can break things/rename/revert stuff etc without pissing off folks :)

Any opnions?

BulatSaif commented:
> Another option is call this option `--rpc-endpoint-unstable` until we collect some feedback on this approach and can break things/rename/revert stuff etc without pissing off folks :)

As I've seen in various applications, the common practice is to place `--rpc-endpoint` in the Experimental flags section in `--help`. If you want to clearly mark the flag, it's better to use the term "experimental" rather than "unstable."

skunert commented:
> > Another option is call this option `--rpc-endpoint-unstable` until we collect some feedback on this approach and can break things/rename/revert stuff etc without pissing off folks :)
> 
> As I've seen in various applications, the common practice is to place `--rpc-endpoint` in the Experimental flags section in `--help`. If you want to clearly mark the flag, it's better to use the term `experimental` rather than `unstable`.

Yes, following the pattern of our other experimental flags, it would be `--experimental-rpc-endpoints`.

niklasad1 commented on substrate/client/cli/src/commands/run_cmd.rs:
```diff
@@ -128,11 +128,12 @@ pub struct RunCmd {
-	/// The format for this option is: `--rpc-endpoint"listen-addr=<ip:port>,<key=value>,..."`
-	/// where each option is separated by a comma and `listen-addr` is the only required param.
+	/// The format for this option is:
+	/// `--experimental-rpc-endpoint"listen-addr=<ip:port>,<key=value>,..."` where each option is
```
```suggestion
	/// `--experimental-rpc-endpoint" listen-addr=<ip:port>,<key=value>,..."` where each option is
```

lexnv reviewed (approved):
LGTM! Nice job here 🙏

niklasad1 commented on substrate/client/cli/src/commands/run_cmd.rs:
```diff
@@ -128,6 +128,46 @@ pub struct RunCmd {
+	///    (optional)
+	///  • rate-limit: The rate limit in calls per minute for each connection (optional)
+	///  • rate-limit-trust-proxy-headers: Trust proxy headers for disable rate limiting (optional)
+	///  • rate-limit-whitelisted-ips: Disable rate limiting for certain ip addresses (optional)
```
```suggestion
	///  • rate-limit-whitelisted-ips: Disable rate limiting for certain ip addresses, this can be enabled more than once (optional)
```

niklasad1 commented on substrate/client/rpc-servers/src/lib.rs:
```diff
@@ -23,246 +23,255 @@
+					.custom_tokio_runtime(cfg.tokio_handle.clone())
+					.set_id_provider(RandomStringIdProvider::new(16));
+
+				if let Some(provider) = id_provider2.take() {
```
Hey, if you take a look we are now using a `tokio::spawn task` for each RPC endpoint and that's why we need `take() and clone()` because it's used in a loop

TarekkMA replied:
Thank you for the response,

In [a test in moonbeam](https://github.com/moonbeam-foundation/moonbeam/blob/c6d07d703dfcdd94cc311fa83b553071b7d433ff/test/suites/dev/moonbase/test-subscription/test-subscription.ts#L20-L31) we found that the `id_provider` is being reset somehow  and changed to the default one. Changing `.take()` to `.clone()` would fix the issue.

I've opened a PR: https://github.com/paritytech/polkadot-sdk/pull/6588 addressing this issue

niklasad1 replied:
Ok, weird that wasn't obvious to me but your fix looks good to me

Polkadot-Forum commented:
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/substrate-rpc-server-cli-changes-updates/11122/1
